### PR TITLE
Multiple improvements:

### PR DESCRIPTION
**Pull request by marcuswestin**
- Make `dalai llama` script idempotent:
    1: Don't re-download files that have already been downloaded
    2: Don't re-create models that have already been created
- Allow for specifying which model to run in the UI
    - show an error message if the chosen model hasn't been downloaded and created yet
    - and show instructions for how to do download and create it if it doesn't exist (`dalai llama <MODEL>`)
- Better log statements during download, model creation, and request serving
- Create yarn scripts `yarn dalai:llama <optional MODEL>` and `yarn start`
- Put the dalai llama directory in `cwd` by default, and allow for specifying where to put it with the `LLAMA_DIR` environment variable
- Remove express url encoded warning message on server startup by making the extended option explicit (see https://github.com/expressjs/body-parser#extended for more info)
- Allow for the server to optionally end with a "\n\n<end>" message when the response is finished
- Add `./dalai` script in dalai repo to allow for easier local development testing of the CLI
- Exit with success code 0 when `dalai llama` succeeds

### CODE REVIEW

**Comment by marcuswestin** [comment]
Thanks @fermigas 

This patch should be a meaningful improvement for everyone that is playing with this I think

@cocktailpeanut would love a code review and merge, or pointers for what to update :)

File: bin/web/index.js [hunk 1]
```diff
@@ -9,14 +9,14 @@ const start = (port) => {
   dalai.http(httpServer)
   app.use(express.static(path.resolve(__dirname, 'public')))
   app.use(express.json());
-  app.use(express.urlencoded());
+  app.use(express.urlencoded({ extended: true }));
```
Comment (fermigas) [inline bin/web/index.js:12]: Good.  Fixes bug where server emits this line on stating:  

body-parser deprecated undefined extended: provide extended option .npm/_npx/3c737cbb02d79cc9/node_modules/dalai/bin/web/index.js:12:19

File: bin/web/views/index.ejs [hunk 2]
```diff
@@ -30,6 +30,13 @@
 </head>
 <body>
 <form id="form" action="">
+  <select id="model">
+    <!-- options: 7B, 13B, 30B, 65B -->
+    <option value="7B">7B</option>
+    <option value="13B">13B</option>
+    <option value="30B">30B</option>
+    <option value="65B">65B</option>
+  </select>
```
Comment (fermigas) [inline bin/web/views/index.ejs:39]: Really useful, and works fine.  
Might consider adding a  text input box for # of threads, and temperature too.   I'm finding the model pretty sensitive to temperature settings.

File: index.js [hunk 3]
```diff
@@ -36,6 +37,11 @@ class Dalai {
     await fs.promises.mkdir(resolvedPath, { recursive: true }).catch((e) => { })
 
     for(let file of files) {
+      if (fs.existsSync(path.resolve(resolvedPath, file))) {
+        console.log(`Skip file download, it already exists: ${file}`)
+        continue;
+      }
+
```
Comment (fermigas) [inline index.js:44]: Very useful.  The 6th parameter file of the 65B download kept failing for me.  I had to pull it from a torrent to get it to work, then had to do the quantization manually.   This and the quantization file check below improve error recover a lot.

File: index.js [hunk 4]
```diff
@@ -117,12 +132,19 @@ class Dalai {
     }
   }
   async query(req, cb) {
+    console.log(`> query:`, req)
```
Comment (fermigas) [inline index.js:135]: Very useful.  You might consider logging the model's output to the console, too.

File: index.js [hunk 5]
```diff
@@ -188,7 +215,13 @@ class Dalai {
...
     for(let i=0; i<num[model]; i++) {
       const suffix = (i === 0 ? "" : `.${i}`)
-      await this.exec(`./quantize ./models/${model}/ggml-model-f16.bin${suffix} ./models/${model}/ggml-model-q4_0.bin${suffix} 2`, this.home)
+      const outputFile1 = `./models/${model}/ggml-model-f16.bin${suffix}`
+      const outputFile2 = `./models/${model}/ggml-model-q4_0.bin${suffix}`
+      if (fs.existsSync(path.resolve(this.home, outputFile1)) && fs.existsSync(path.resolve(this.home, outputFile2))) {
+        console.log(`Skip quantization, files already exists: ${outputFile1} and ${outputFile2}}`)
+        continue
+      }
+      await this.exec(`./quantize ${outputFile1} ${outputFile2} 2`, this.home)
```
Comment (fermigas) [inline index.js:224]: As mentioned above, this is a big improvement.

**Review by fermigas — COMMENTED** [review]
I manually made the changes I reviewed below and tested them.  
Downloads and quantization code changes work fine.  Approved.
All 4 models work with the new UI changes.   Approved.  

I did not apply the changes in the files I haven't marked "viewed".

**Comment by marcuswestin** [comment]
👍 🙏

**Pull request merged**